### ISSUE TITLE
unique_copy + clang: Invalid __global__ read of size 4, code later crashes

### ISSUE THREAD

**Issue by andrewcorrigan** [NVIDIA/thrust#865]
When running my code, which uses Thrust (latest git revision), using the Clang/CUDA compiler (latest revision as of Nov. 8), a call to `thrust::unique_copy` with a particular data set on a particular (important) code path leads to memory access errors, as reported by `cuda-memcheck`.   Interestingly, it does compute the result correctly, and earlier calls to `thrust::unique_copy` do not trigger memory access errors.  Despite computing the correct output, later on my code crashes, and I suspect that this earlier invalid memory access is the cause.

A minimal reproducer is below, with the problematic data set embedded inside.  I compiled with CUDA 8 and sm_35, but the invalid memory access occurs for other CUDA versions (I also tried 7.0) and architectures (I also tried 2.0 and 5.2).

@jlebar is unable to rule out [this known ptxas issue](https://llvm.org/bugs/show_bug.cgi?id=27738).  I was wondering if there might still be a Thrust bug, or at least that a workaround might be possible, specifically regarding this:
```
In fact one of the thrust [0] testcases fails at O2 with clang, apparently
due to this bug (thrust uses the "if thread 0 write some state, then
syncthreads" idiom in many places). 
```

Minimal reproducer:
```
#include <string>
#include <sstream>
#include <iostream>

#define THRUST_DEBUG 1
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>
#include <thrust/unique.h>

int main(int argc, char** argv)
{
    std::stringstream f(std::string("0 0 1 1 1 2 2 2 2 2 2 4 4 4 5 5 5 5 5 5 7 7 7 8 8 8 8 8 8 10 10 10 11 11 11 11 11 11 13 13 13 14 14 14 14 14 14 16 16 16 17 17 17 17 17 17 19 19 19 20 20 20 20 20 20 22 22 22 23 23 23 23 23 23 25 25 25 26 26 26 26 26 26 28 29 29 29 30 30 30 32 32 32 32 32 32 35 35 35 35 35 35 38 38 38 38 38 38 41 41 41 41 41 41 44 44 44 44 44 44 47 47 47 47 47 47 50 50 50 50 50 50 53 53 53 53 53 53 56 56 56 56 56 56 59 59 59 60 60 60 62 62 62 62 62 62 65 65 65 65 65 65 68 68 68 68 68 68 71 71 71 71 71 71 74 74 74 74 74 74 77 77 77 77 77 77 80 80 80 80 80 80 83 83 83 83 83 83 86 86 86 86 86 86 89 89 89 90 90 90 92 92 92 92 92 92 95 95 95 95 95 95 98 98 98 98 98 98 101 101 101 101 101 101 104 104 104 104 104 104 107 107 107 107 107 107 110 110 110 110 110 110 113 113 113 113 113 113 116 116 116 116 116 116 119 119 119 120 120 120 122 122 122 125 125 125 128 128 128 131 131 131 134 134 134 137 137 137 140 140 140 143 143 143 146 146 146 149 149 272"));
    thrust::host_vector<int> h_data;
    while(f.good())
    {
        int i;
        f >> i;
        h_data.push_back(i);
    }
    if(h_data.size() != 300)
    {
        throw std::runtime_error("Unexpected size");
    }

    thrust::device_vector<int> data(h_data);
    thrust::device_vector<int> data_unique(data.size());

    int n_unique =
    thrust::unique_copy(data.begin(), data.end(), data_unique.begin()) - data_unique.begin();


    if(n_unique != 66)
    {
        throw std::runtime_error("Unexpected number of unique values");
    }

    return 0;
 }
```
With `nvcc`:
```
nvcc  -O3 -gencode arch=compute_35,code=sm_35 unique_copy.cu -I/path/to/thrust
cuda-memcheck ./a.out 
========= CUDA-MEMCHECK
========= ERROR SUMMARY: 0 errors
```

With `clang`:
```
clang++ -x cuda  -stdlib=libc++ -O3 --cuda-gpu-arch=sm_35 --cuda-path=/usr/local/cuda-8.0 unique_copy.cu -I/path/to/thrust -L/usr/local/cuda-8.0/lib64 -lcudart_static -lpthread  -ldl -lrt
cuda-memcheck ./a.out 
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 4
=========     at 0x000001e8 in void thrust::system::cuda::detail::detail::launch_closure_by_value<thrust::system::cuda::detail::commutative_reduce_intervals_closure<thrust::transform_iterator<thrust::detail::predicate_to_integral<thrust::detail::functional::actor<thrust::detail::functional::argument<unsigned int=0>>, long>, thrust::transform_iterator<thrust::detail::head_flags<thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::equal_to<int>, bool, long>::head_flag_functor, thrust::zip_iterator<thrust::tuple<thrust::counting_iterator<long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::detail::normal_iterator<thrust::device_ptr<int>>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, thrust::use_default, thrust::use_default>, long, thrust::use_default>, thrust::detail::normal_iterator<thrust::pointer<long, thrust::system::cuda::detail::tag, thrust::use_default
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0xb0573fffc is out of bounds
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib64/libcuda.so.1 (cuLaunchKernel + 0x2c5) [0x203f65]
=========     Host Frame:./a.out [0x23bb1]
=========     Host Frame:./a.out [0x41753]
=========     Host Frame:./a.out [0xaa1f]
=========     Host Frame:./a.out [0xa57c]
=========     Host Frame:./a.out [0x4e34]
=========     Host Frame:/lib64/libc.so.6 (__libc_start_main + 0xf5) [0x21af5]
=========     Host Frame:./a.out [0x4aed]
=========
========= Program hit cudaErrorLaunchFailure (error 4) due to "unspecified launch failure" on CUDA API call to cudaDeviceSynchronize. 
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib64/libcuda.so.1 [0x2eea03]
=========     Host Frame:./a.out [0x41476]
=========     Host Frame:./a.out [0xb2ec]
=========     Host Frame:./a.out [0xaa29]
=========     Host Frame:./a.out [0xa57c]
=========     Host Frame:./a.out [0x4e34]
=========     Host Frame:/lib64/libc.so.6 (__libc_start_main + 0xf5) [0x21af5]
=========     Host Frame:./a.out [0x4aed]
=========
terminating with uncaught exception of type thrust::system::system_error: launch_closure_by_value: unspecified launch failure
========= Error: process didn't terminate successfully
========= Internal error (20)
========= No CUDA-MEMCHECK results found
```

**Comment by brycelelbach** [comment]
What linux distro and nvidia kernel driver version?

**Comment by andrewcorrigan** [comment]
This was CentOS 7.  I'll reproduce it on Monday and check the driver version.

**Comment by brycelelbach** [comment]
Sorry, I should've mentioned it was gonna take me a few days to get to this.

The version of Thrust in CUDA 9.0 (1.9.0) has a new CUDA backend, so this ptxas bug may no longer be triggered. I know that @jlebar and co are working on CUDA 9.0 support, but I'm not sure what state it's in.

Once it's available, we can test against it and go from there.

**Comment by jlebar** [comment]
Clang currently has incomplete support for CUDA 9 -- it works, but not (yet) if you include some of the new headers.  I don't think we've tested Thrust specifically.

But if clang doesn't yet build Thrust with CUDA 9, you can just substitute ptxas9 into a cuda8 build directory.  That should work fine.

**Comment by brycelelbach** [comment]
Hey @andrewcorrigan, is this still a problem?

**Comment by andrewcorrigan** [comment]
Sorry for not responding, I haven't encountered this issue for quite a while.
